### PR TITLE
Integrate testing into cargo test

### PR DESCRIPTION
**Pull request by JackWolfard**
To address #56, this PR embeds testing logic directly in Rust and is meant to replicate the existing functionality in `ci/selftest.sh`.

However, I discovered a bug with `--format` (#57) which prevented me from adequately covering the `--format` and `--prefix` tests in `ci/selftest.sh`.

To run the new tests, just use `cargo test`. Alternatively, you can specify `cargo test --test integration` to filter just for the tests I have added.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -32,3 +32,6 @@ either = "1.7.0"
 
 [features]
 vendored-openssl = ["openssl/vendored"]
+
+[dev-dependencies]
+once_cell = "1.17.1"
```
Comment (JackWolfard) [inline Cargo.toml:37]: No longer needed. Can remove.

File: tests/vendor_filterer/common.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        // io::stdout().write_all(&output.stdout)?;
+        // io::stderr().write_all(&output.stderr)?;
+        // use std::{thread, time};
+        // thread::sleep(time::Duration::from_millis(200));
+        output
+    })
+}
+
+pub fn dir() -> Result<Utf8PathBuf> {
+    // to ensure each call to dir gets a unique directory, use atomics
```
Comment (cgwalters) [inline tests/vendor_filterer/common.rs:130]: Why not `tempfile::tempdir()`?

File: tests/vendor_filterer/common.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            format: None,
+            manifest_path: None,
+        }
+    }
+}
+
+pub fn vendor(options: VendorOptions) -> Result<Output> {
+    use once_cell::sync::OnceCell;
+    use std::sync::Mutex;
+    static LOCK: OnceCell<Mutex<()>> = OnceCell::new();
```
Comment (cgwalters) [inline tests/vendor_filterer/common.rs:89]: The rationale for this is to limit concurrency?  If so, why?  Stuff like this really wants a code comment.

**Review by cgwalters — COMMENTED** [review]
Gave this a read; looks sane, thanks!  Are you thinking merging this blocks on not forking off external executables?

We could probably for now only conditionally run these tests on Linux I guess?

**Comment by cgwalters** [comment]
OK I've made some minor cleanups in this PR, and also rebased it on top of https://github.com/coreos/cargo-vendor-filterer/pull/64

- Use `tempdir()` in tests
- Prefer `pub(crate)` just on principle
- Use `std::fs::write` instead of inlining it
- Add a few doc comments

And I also took the step of deleting the shell script testing.

Thanks for this PR!

**Review by cgwalters — APPROVED** [review]

**Pull request merged**